### ISSUE TITLE
[Bug] STM32 Serial Mavlink on Main Disconnects

### ISSUE THREAD

**Issue by AlexKlimaj** [PX4/PX4-Autopilot#22599]
### Describe the bug

On current main with the recent DMA changes, I cannot keep serial mavlink connected to a companion computer. It will run for a few minutes before locking up and needing a reboot of the flight controller. With the DMA changes reverted, it stays connected indefinitely. 

### To Reproduce

Run mavlink serial on main with an STM32H7 connected directly to a companion computer.

### Expected behavior

Stays connected. 

### Flight Log

NA

### Software Version

main after nuttx st dma changes

### Flight controller

ARKV6X

### Vehicle type

None

**Comment by dagar** [comment]
@davids5 @niklaut @AlexKlimaj can we hop on a call to discuss this as soon as possible? If mavlink is broken on STM32H7 in main this is pretty serious.

**Comment by davids5** [comment]
> @davids5 @niklaut @AlexKlimaj can we hop on a call to discuss this as soon as possible? If mavlink is broken on STM32H7 in main this is pretty serious.

@dagar Sure - let's coordinate on `discord maintainers`  I have openings 8-12 EST today

**Comment by AlexKlimaj** [comment]
I think we need to revert these commits and re-evaluate them.

https://github.com/PX4/NuttX/commit/ed4814f6239097dde5eecf2b4fbd58661db84dda
https://github.com/PX4/NuttX/commit/3dc3cf522758d70ca2a07c156e5e02517650327d

**Comment by davids5** [comment]
No @AlexKlimaj.  A non blocking operations should not block. Period! Test it. Pull CTS false the system will hang!  Wrong is wrong! Let's Fix the root cause.

**Comment by niklaut** [comment]
I'm confused that TELEM1,2,3 are all configured with flow control in NuttX, but Mavlink status only shows the flow control for TELEM3. Flow control is also off when it's working? Is the flow control even physically used by the other communication side?

**Comment by AlexKlimaj** [comment]
I tested with and without flow control connected. The lock up occurs in both cases.

**Comment by davids5** [comment]
@AlexKlimaj Did you terminate the CTS Lpw?

**Comment by AlexKlimaj** [comment]
No it was floating. But the mavlink driver is disabling flow control. Shouldn't the state of the pin not matter in that instance?

**Comment by davids5** [comment]
Should be is the Key word here.

Please get a scope on the Serial lines and HS and let's see what your CPU is seeing

**Comment by niklaut** [comment]
Also, [looking at the code, hardware flow control can automatically be disabled in `FLOW_CONTROL_AUTO` mode?](https://github.com/PX4/PX4-Autopilot/blob/20129e63facab129ad31fa693d1397e7458afaa4/src/modules/mavlink/mavlink_main.cpp#L730). The flow control mode is controlled by the `-z` (on), `-Z` (off)`, or none (auto) flags of `mavlink start`. So it may start with flow control on, but then gets switched off at some time…

**Comment by AlexKlimaj** [comment]
Yes but this lock up occurs on mavlink and dds. dds always has flow control turned off. Serial TX will run for a short period of time then just stop outputting.

**Comment by davids5** [comment]
Off and on are time relative. At boot up it is on until turned off. If there is trapped data when it was on, that could cause the issues. Please try to pull the line Low from boot and see if it affects it.

**Comment by AlexKlimaj** [comment]
https://github.com/PX4/PX4-Autopilot/pull/22614/files

With serial TX DMA disabled, the DDS client is still working 9 hours later.

**Comment by davids5** [comment]
@AlexKlimaj @niklaut has been trying to reproduce the failure you are seeing. But can not repro it yet. Do you have a Salea Logic analyzer? If So I can add Probes to the Serial driver that you can run in the failing system. The LA is connected to the PWM OUT pins. This way we can see the state and see if there is a race we are not seeing.

**Comment by niklaut** [comment]
I've been running the serial_test for several hours and cannot reproduce this on our Skynode X config (FMUv6x with the same config as yours). 

```
/dev/ttymxc3: count for this session: rx=579016685, tx=571762328, rx err=6024
/dev/ttymxc3: TIOCGICOUNT: ret=0, rx=579107680, tx=571859587, frame = 0, overrun = 0, parity = 0, brk = 0, buf_overrun = 0
```

I'm also working on understanding max throughput better with the serial test via tracing. However, DMA utilization seems very high on first glance (we use a 3Mbps UART link to the companion). I'll investigate further. [orbetto.perf.gz -> ui.perfetto.io](https://github.com/PX4/PX4-Autopilot/files/13891416/orbetto.perf.gz)

**Comment by davids5** [comment]
@niklaut What about just using mavlink?

**Comment by PetervdPerk-NXP** [comment]
@niklaut you've tested asymmetrical flow? For example only tx is transmitting and rx doesn't send anything. I've got a hunch the issues come from the fact that mavlink and xrce don't constantly flood both tx rx like the serial test does.

**Comment by niklaut** [comment]
I've tried a number of things including running serial test asymmetrically. I just cannot reproduce this and I haven't gotten any reports of this failing in the field in our test vehicles. I'll revisit this once I can get a handle on it.

**Comment by AlexKlimaj** [comment]
Just got a binary built on GCC 9.3.1 to fail. Testing the same firmware built with GCC 10.2.1 now. Have a scope hooked up to TX/RX and I'm waiting to see if it fails. When it does fail, TX is idling high and QGC is sending the heartbeat.

**Comment by Igor-Misic** [comment]
Here is one of the way to reproduce this issue relatively quickly.
- start mavlink on one of the UARTs with enabled **TX DMA** (_example: mavlink start -d /dev/ttyS2 -b 921600 -r 290000 -m onboard_low_bandwidth -x_) 
- connect QGC to FMU over previously set UART
- start downloading one log file with 200 MB or more and wait

In 90% of time I wasn't able to finish download.

**Comment by davids5** [comment]
> Here is one of the way to reproduce this issue relatively quickly.
> 
> * start mavlink on one of the UARTs with enabled **TX DMA** (_example: mavlink start -d /dev/ttyS2 -b 921600 -r 290000 -m onboard_low_bandwidth -x_)
> * connect QGC to FMU over previously set UART
> * start downloading one log file with 200 MB or more and wait
> 
> In 90% of time I wasn't able to finish download. In my case, I don't have flow control enabled nor does it exist on hardware.

@Igor-Misic What HW are you testing on and what is connected?

**Comment by Igor-Misic** [comment]
> @Igor-Misic What HW are you testing on and what is connected?

Unfortunately, it is a development board with STM32H7 without USB so I was forced to use UART to connect QGC. 
Unresponsive/blocked UART led me to this thread. 
Maybe my issue is not the same as this one but it seems I can't reproduce it if I don't use Tx DMA.  Hm...

**Comment by davids5** [comment]
I have added PROBEs instrumentation on the V6X. N.B the PWM is disabled and used for PROBES.

https://github.com/PX4/PX4-Autopilot/tree/main-dnm-probes-h7-serial

The Second commit - may be the solution. Can you test this branch?

**Comment by Igor-Misic** [comment]
> The Second commit - may be the solution. Can you test this branch?

So far it works!

**Comment by davids5** [comment]
@AlexKlimaj @Igor-Misic - Thank you both for testing.

After sleeping on it, I realized there was still a window that could be interrupted. It is much smaller, then it used to be, but still exited. 

I have closed the window. Please test main using [this NuttX](https://github.com/PX4/NuttX/tree/px4_firmware_nuttx-10.3.0%2B-pr-h7-serial-dma-busy)

**Comment by davids5** [comment]
Here is what was happening:

Normal

![normal](https://github.com/PX4/PX4-Autopilot/assets/1945821/11dbfdb6-711c-44dd-9b20-911dab2434d2)

Caused the hang


![hang](https://github.com/PX4/PX4-Autopilot/assets/1945821/ac7482d6-e2df-468a-b4e3-ed6ce18b5a9d)

**Comment by davids5** [comment]
@Igor-Misic & @AlexKlimaj did you test 56aa08568f8eb84f8096114e1a9485fb0d3b0d18 ?

**Comment by Igor-Misic** [comment]
I have applied your patch and since then haven't seen any issues. Also, it works ~20% faster than previously.

**Comment by niklaut** [comment]
I'm soak testing this change on our CI for the next 12h, so we'll know if it impacts our setup too.

**Comment by AlexKlimaj** [comment]
Looking good!

**Comment by niklaut** [comment]
There's no change in behavior in our CI. Which goes in line with my inability to reproduce this bug on our hardware setup. So… merge?

**Comment by MaEtUgR** [comment]
Fix https://github.com/PX4/PX4-Autopilot/pull/22667 was merged, please reopen if the problem persists.
